### ISSUE TITLE
[Feature] Optimize parquet nested col read performance 

### ISSUE THREAD

**Issue by Zouxxyy** [apache/paimon#4447]
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/paimon/issues) and found nothing similar.

### Motivation

NestedColumnReader has performance regression for simple row col reading

using NestedColumnReader

<img width="1415" alt="image" src="https://github.com/user-attachments/assets/4835249c-5d61-40f6-a703-15bd1e25f002">

using RowColumnReader

<img width="1417" alt="image" src="https://github.com/user-attachments/assets/77c56700-98ba-4c08-84e9-b4a27346bae3">

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

**Comment by JingsongLi** [comment]
@Zouxxyy You are correct, yes, now this implementation has some performance regression.
